File: backend/controllers/theThanhVienController.js
const TheThanhVien = require("../models/TheThanhVien");
const TaiKhoan = require("../models/TaiKhoan");
const { Op } = require("sequelize");

const validateForm = ({
  ten_hang,
  ty_le_giam_gia,
  muc_chi_tieu_tu,
  diem_thuong_them,
}) => {
  if (!ten_hang?.trim()) return "Tên hạng thẻ không được để trống!";
  if (muc_chi_tieu_tu < 0) return "Mức chi tiêu không được là số âm!";
  if (ty_le_giam_gia < 0 || ty_le_giam_gia > 100)
    return "Tỷ lệ giảm giá phải từ 0 đến 100%!";
  if (diem_thuong_them < 0) return "Điểm thưởng thêm không được là số âm!";
  return null;
};

exports.getAllTheThanhVien = async (req, res) => {
  try {
    const theThanhVien = await TheThanhVien.findAll({
      order: [["muc_chi_tieu_tu", "ASC"]],
    });
    res.status(200).json(theThanhVien);
  } catch (error) {
    console.error("Lỗi lấy danh sách thẻ thành viên:", error);
    res.status(500).json({ message: "Lỗi server" });
  }
};

exports.createTheThanhVien = async (req, res) => {
  try {
    const {
      ten_hang,
      muc_chi_tieu_tu = 0,
      ty_le_giam_gia = 0,
      diem_thuong_them = 0,
      mau_the = "#2563eb",
      mo_ta_quyen_loi = "",
    } = req.body || {};

    const validErr = validateForm({
      ten_hang,
      ty_le_giam_gia,
      muc_chi_tieu_tu,
      diem_thuong_them,
    });
    if (validErr) return res.status(400).json({ message: validErr });

    const existing = await TheThanhVien.findOne({
      where: { ten_hang: ten_hang.trim() },
    });
    if (existing)
      return res
        .status(400)
        .json({ message: `Hạng thẻ "${ten_hang}" đã tồn tại!` });

    const newThe = await TheThanhVien.create({
      ten_hang: ten_hang.trim(),
      muc_chi_tieu_tu: Number(muc_chi_tieu_tu),
      ty_le_giam_gia: Number(ty_le_giam_gia),
      diem_thuong_them: Number(diem_thuong_them),
      mau_the,
      mo_ta_quyen_loi,
    });

    res
      .status(201)
      .json({ message: "Thêm hạng thẻ thành công!", data: newThe });
  } catch (error) {
    console.error("Lỗi thêm thẻ:", error);
    res.status(500).json({ message: "Lỗi khi thêm hạng thẻ" });
  }
};

exports.updateTheThanhVien = async (req, res) => {
  try {
    const { id } = req.params;
    const {
      ten_hang,
      muc_chi_tieu_tu,
      ty_le_giam_gia,
      diem_thuong_them,
      mau_the,
      mo_ta_quyen_loi,
    } = req.body || {};

    const the = await TheThanhVien.findByPk(id);
    if (!the)
      return res.status(404).json({ message: "Không tìm thấy hạng thẻ!" });

    const toValidate = {
      ten_hang: ten_hang ?? the.ten_hang,
      ty_le_giam_gia: ty_le_giam_gia ?? the.ty_le_giam_gia,
      muc_chi_tieu_tu: muc_chi_tieu_tu ?? the.muc_chi_tieu_tu,
      diem_thuong_them: diem_thuong_them ?? the.diem_thuong_them,
    };
    const validErr = validateForm(toValidate);
    if (validErr) return res.status(400).json({ message: validErr });

    if (ten_hang && ten_hang.trim() !== the.ten_hang) {
      const duplicate = await TheThanhVien.findOne({
        where: {
          ten_hang: ten_hang.trim(),
          id: { [Op.ne]: id },
        },
      });
      if (duplicate)
        return res
          .status(400)
          .json({ message: `Hạng thẻ "${ten_hang}" đã tồn tại!` });
    }

    await the.update({
      ten_hang: ten_hang !== undefined ? ten_hang.trim() : the.ten_hang,
      muc_chi_tieu_tu:
        muc_chi_tieu_tu !== undefined
          ? Number(muc_chi_tieu_tu)
          : the.muc_chi_tieu_tu,
      ty_le_giam_gia:
        ty_le_giam_gia !== undefined
          ? Number(ty_le_giam_gia)
          : the.ty_le_giam_gia,
      diem_thuong_them:
        diem_thuong_them !== undefined
          ? Number(diem_thuong_them)
          : the.diem_thuong_them,
      mau_the: mau_the !== undefined ? mau_the : the.mau_the,
      mo_ta_quyen_loi:
        mo_ta_quyen_loi !== undefined ? mo_ta_quyen_loi : the.mo_ta_quyen_loi,
    });

    res.status(200).json({ message: "Cập nhật thành công!", data: the });
  } catch (error) {
    console.error("Lỗi cập nhật thẻ:", error);
    res.status(500).json({ message: "Lỗi cập nhật hạng thẻ" });
  }
};

exports.deleteTheThanhVien = async (req, res) => {
  try {
    const the = await TheThanhVien.findByPk(req.params.id);
    if (!the)
      return res.status(404).json({ message: "Không tìm thấy hạng thẻ!" });

    // Kiểm tra có user nào đang ở hạng này không
    const soLuongUser = await TaiKhoan.count({
      where: { the_thanh_vien_id: req.params.id },
    });
    if (soLuongUser > 0) {
      return res.status(400).json({
        message: `Không thể xóa! Hiện có ${soLuongUser} khách hàng đang ở hạng "${the.ten_hang}". Vui lòng chuyển họ sang hạng khác trước.`,
      });
    }

    await the.destroy();
    res.status(200).json({ message: "Xóa hạng thẻ thành công!" });
  } catch (error) {
    console.error("Lỗi xóa thẻ:", error);
    res.status(500).json({ message: "Lỗi xóa hạng thẻ" });
  }
};

File: backend-csharp/Controllers/TheThanhVienController.cs
using System;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using WebFashion.Api.Models;

namespace WebFashion.Api.Controllers
{
    [Route("api/memberships")]
    [ApiController]
    [Authorize]
    public class TheThanhVienController : ControllerBase
    {
        private readonly AppDbContext _context;

        public TheThanhVienController(AppDbContext context)
        {
            _context = context;
        }

        private bool IsAdmin()
        {
            var userRole = User.FindFirst("vai_tro")?.Value;
            return userRole == "admin";
        }

        private string? ValidateForm(string? tenHang, decimal tyLeGiamGia, decimal mucChiTieuTu, int diemThuongThem)
        {
            if (string.IsNullOrWhiteSpace(tenHang)) return "Tên hạng thẻ không được để trống!";
            if (mucChiTieuTu < 0) return "Mức chi tiêu không được là số âm!";
            if (tyLeGiamGia < 0 || tyLeGiamGia > 100) return "Tỷ lệ giảm giá phải từ 0 đến 100%!";
            if (diemThuongThem < 0) return "Điểm thưởng thêm không được là số âm!";
            return null;
        }

        // 1. LẤY DANH SÁCH HẠNG THẺ (Dành cho Admin)
        // GET: api/memberships
        [HttpGet]
        public async Task<IActionResult> GetAllTheThanhVien()
        {
            try
            {
                if (!IsAdmin())
                {
                    return StatusCode(403, new { message = "Quyền truy cập bị từ chối!" });
                }

                var list = await _context.TheThanhViens
                    .OrderBy(m => m.MucChiTieuTu)
                    .ToListAsync();

                return Ok(list);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi lấy danh sách thẻ thành viên: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi server" });
            }
        }

        // 2. THÊM HẠNG THẺ MỚI
        // POST: api/memberships
        [HttpPost]
        public async Task<IActionResult> CreateTheThanhVien([FromBody] JsonElement body)
        {
            try
            {
                if (!IsAdmin())
                {
                    return StatusCode(403, new { message = "Quyền truy cập bị từ chối!" });
                }

                string? tenHang = body.TryGetProperty("ten_hang", out var tenProp) ? tenProp.GetString() : null;
                decimal mucChiTieuTu = body.TryGetProperty("muc_chi_tieu_tu", out var tuProp) ? tuProp.GetDecimal() : 0m;
                decimal tyLeGiamGia = body.TryGetProperty("ty_le_giam_gia", out var giamProp) ? giamProp.GetDecimal() : 0m;
                int diemThuongThem = body.TryGetProperty("diem_thuong_them", out var diemProp) ? diemProp.GetInt32() : 0;
                string mauThe = body.TryGetProperty("mau_the", out var mauProp) ? (mauProp.GetString() ?? "#2563eb") : "#2563eb";
                string moTaQuyenLoi = body.TryGetProperty("mo_ta_quyen_loi", out var moTaProp) ? (moTaProp.GetString() ?? "") : "";

                var validErr = ValidateForm(tenHang, tyLeGiamGia, mucChiTieuTu, diemThuongThem);
                if (validErr != null)
                {
                    return BadRequest(new { message = validErr });
                }

                string trimmedTenHang = tenHang!.Trim();
                var existing = await _context.TheThanhViens.AnyAsync(m => m.TenHang == trimmedTenHang);
                if (existing)
                {
                    return BadRequest(new { message = $"Hạng thẻ \"{tenHang}\" đã tồn tại!" });
                }

                var newThe = new TheThanhVien
                {
                    TenHang = trimmedTenHang,
                    MucChiTieuTu = mucChiTieuTu,
                    TyLeGiamGia = tyLeGiamGia,
                    DiemThuongThem = diemThuongThem,
                    MauThe = mauThe,
                    MoTaQuyenLoi = moTaQuyenLoi
                };

                _context.TheThanhViens.Add(newThe);
                await _context.SaveChangesAsync();

                return StatusCode(201, new { message = "Thêm hạng thẻ thành công!", data = newThe });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi thêm thẻ thành viên: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi khi thêm hạng thẻ" });
            }
        }

        // 3. CẬP NHẬT CHI TIẾT HẠNG THẺ
        // PUT: api/memberships/{id}
        [HttpPut("{id}")]
        public async Task<IActionResult> UpdateTheThanhVien(long id, [FromBody] JsonElement body)
        {
            try
            {
                if (!IsAdmin())
                {
                    return StatusCode(403, new { message = "Quyền truy cập bị từ chối!" });
                }

                var the = await _context.TheThanhViens.FindAsync(id);
                if (the == null)
                {
                    return NotFound(new { message = "Không tìm thấy hạng thẻ!" });
                }

                string? tenHang = body.TryGetProperty("ten_hang", out var tenProp) ? tenProp.GetString() : the.TenHang;
                decimal mucChiTieuTu = body.TryGetProperty("muc_chi_tieu_tu", out var tuProp) ? tuProp.GetDecimal() : the.MucChiTieuTu;
                decimal tyLeGiamGia = body.TryGetProperty("ty_le_giam_gia", out var giamProp) ? giamProp.GetDecimal() : the.TyLeGiamGia;
                int diemThuongThem = body.TryGetProperty("diem_thuong_them", out var diemProp) ? diemProp.GetInt32() : the.DiemThuongThem;
                string? mauThe = body.TryGetProperty("mau_the", out var mauProp) ? mauProp.GetString() : the.MauThe;
                string? moTaQuyenLoi = body.TryGetProperty("mo_ta_quyen_loi", out var moTaProp) ? moTaProp.GetString() : the.MoTaQuyenLoi;

                var validErr = ValidateForm(tenHang, tyLeGiamGia, mucChiTieuTu, diemThuongThem);
                if (validErr != null)
                {
                    return BadRequest(new { message = validErr });
                }

                if (tenHang != null && tenHang.Trim() != the.TenHang)
                {
                    string trimmedTen = tenHang.Trim();
                    var duplicate = await _context.TheThanhViens.AnyAsync(m => m.TenHang == trimmedTen && m.Id != id);
                    if (duplicate)
                    {
                        return BadRequest(new { message = $"Hạng thẻ \"{tenHang}\" đã tồn tại!" });
                    }
                    the.TenHang = trimmedTen;
                }

                the.MucChiTieuTu = mucChiTieuTu;
                the.TyLeGiamGia = tyLeGiamGia;
                the.DiemThuongThem = diemThuongThem;
                if (mauThe != null) the.MauThe = mauThe;
                if (moTaQuyenLoi != null) the.MoTaQuyenLoi = moTaQuyenLoi;

                await _context.SaveChangesAsync();

                return Ok(new { message = "Cập nhật thành công!", data = the });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi cập nhật thẻ thành viên: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi cập nhật hạng thẻ" });
            }
        }

        // 4. XÓA HẠNG THẺ
        // DELETE: api/memberships/{id}
        [HttpDelete("{id}")]
        public async Task<IActionResult> DeleteTheThanhVien(long id)
        {
            try
            {
                if (!IsAdmin())
                {
                    return StatusCode(403, new { message = "Quyền truy cập bị từ chối!" });
                }

                var the = await _context.TheThanhViens.FindAsync(id);
                if (the == null)
                {
                    return NotFound(new { message = "Không tìm thấy hạng thẻ!" });
                }

                // Check if any users are currently in this tier
                int userCount = await _context.TaiKhoans.CountAsync(u => u.TheThanhVienId == id);
                if (userCount > 0)
                {
                    return BadRequest(new
                    {
                        message = $"Không thể xóa! Hiện có {userCount} khách hàng đang ở hạng \"{the.TenHang}\". Vui lòng chuyển họ sang hạng khác trước."
                    });
                }

                _context.TheThanhViens.Remove(the);
                await _context.SaveChangesAsync();

                return Ok(new { message = "Xóa hạng thẻ thành công!" });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi xóa thẻ thành viên: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi xóa hạng thẻ" });
            }
        }
    }
}
